### DISCUSSION TITLE
Why call the getlogs API during hardhat verify?

### DISCUSSION THREAD

**Discussion by kkeolmusae** [NomicFoundation/hardhat#5140] [Troubleshooting]
During hardhat verify, hardhat call `/api?module=logs&action=getLogs&fromBlock=0&toBlock=latest&address=0x55fa02522aDBAc6704e8667859fEa4132D324a48&topic0=0x7e644d79422f17c01e4894b5f4f588d331ebfa28653d42ae832dc59e38c9798f&apikey=unnecessary` 

They call the same api as and can you tell me why? 

Attempted to verify proxy contact.

**Comment by schaable** [comment]
Can you provide an example of where this is happening? I suspect there might be something else going on in your setup, as `hardhat-verify` does not make calls to `getLogs`:

https://github.com/search?q=repo%3ANomicFoundation%2Fhardhat+path%3A%2F%5Epackages%5C%2Fhardhat-verify%5C%2F%2F+getLogs&type=code

**Reply by kkeolmusae** [reply]
First of all, I would like to verify on an explorer called klaytnscope in the klaytn chain. 

https://klaytn.foundation/verifying-contracts-using-hardhat-on-klaytnscope/

The hardhat config is as follows,

```
require("@nomicfoundation/hardhat-toolbox");
require("@openzeppelin/hardhat-upgrades");
require("@nomicfoundation/hardhat-verify");

module.exports = {
  solidity: {
    version: "0.8.18",
    settings: {
      optimizer: {
        enabled: false,
        runs: 200,
      },
    },
  },
  sourcify: {
    enabled: false,
  },
  networks: {
    klaytn: {
      chainId: 8217,
      url: "https://public-en-cypress.klaytn.net",
      // chainId: 1001,
      // url: "https://public-en-baobab.klaytn.net",
    },
  },
  etherscan: {
    apiKey: {
      klaytn: "unnecessary",
    },
    customChains: [
      {
        network: "klaytn",
        // chainId: 1001,
        chainId: 8217,
        urls: {
          apiURL: "https://api-cypress.klaytnscope.com/api",
          browserURL: "https://klaytnscope.com",
          // apiURL: "https://api-baobab.klaytnscope.com/api",
          // browserURL: "https://baobab.klaytnscope.com",
        },
      },
    ],
  },
};

```

When I tried to do hardhat verify, the following error occurred.

```
>> npx hardhat verify --network klaytn 0x55fa02522aDBAc6704e8667859fEa4132D324a48 --contract contracts/Proxy.sol:Proxy 0xa52dEf63b7a3726fF5aF2627E1DCD19476d13B6b
Etherscan:
Failed to get logs for contract at address 0x55fa02522aDBAc6704e8667859fEa4132D324a48.

Etherscan returned with message: NOTOK, reason: Missing Or invalid Module name
```

The following notes the order in which api was requested from hardhat to klaytnscope when performing hardhat verification. There is a part to inquire whether an Admin Changed or upgraded event occurred with getLogs, and after that, it seems that there is a verification for 0x55fa02522aDBAc6704e8667859fEa4132D324a48
When I looked up the [hardhat source code](https://github.com/NomicFoundation/hardhat), I wonder how such api was requested when there was no part calling getLogs in the POST method.

```
0 GET {"apikey":"unnecessary","module":"contract","action":"getsourcecode","address":"0xa52dEf63b7a3726fF5aF2627E1DCD19476d13B6b","page":"1","limit":"25"}

1 POST QueryParams: {"module":"logs","action":"getLogs","fromBlock":"0","toBlock":"latest","address":"0x55fa02522aDBAc6704e8667859fEa4132D324a48","topic0":"0x7e644d79422f17c01e4894b5f4f588d331ebfa28653d42ae832dc59e38c9798f","apikey":"unnecessary","page":"1","limit":"25"}

2 POST QueryParams: {"module":"logs","action":"getLogs","fromBlock":"0","toBlock":"latest","address":"0x55fa02522aDBAc6704e8667859fEa4132D324a48","topic0":"0xbc7cd75a20ee27fd9adebab32041f755214dbc6bffa90cc0225b39da2e5c2d3b","apikey":"unnecessary","page":"1","limit":"25"}

3 GET {"apikey":"unnecessary","module":"contract","action":"getsourcecode","address":"0x55fa02522aDBAc6704e8667859fEa4132D324a48","page":"1","limit":"25"}

4 POST QueryParams: {"module":"contract","action":"verifyproxycontract","address":"0x55fa02522aDBAc6704e8667859fEa4132D324a48","expectedimplementation":"0xa52dEf63b7a3726fF5aF2627E1DCD19476d13B6b","apikey":"unnecessary","page":"1","limit":"25"}
```

For reference, it was confirmed that 'returned with message: NOTOK, reason: Missing Or invalid Module name' occurred because getLogs were not implemented in klaytnscope api. 
I wonder where do hardhat call getLogs and why do call it POST method.

**Reply by schaable** [reply]
I created a new project, deployed, and verified the `Lock` contract with Ignition on Sepolia, setting the `NODE_DEBUG=undici` variable before executing the command. It logged only one POST request, which corresponds to the `verifysourcecode` call.

![image](https://github.com/NomicFoundation/hardhat/assets/1165424/23267c56-f47a-4c2e-8b68-ce7203720943)

I suggest you try again on a fresh Hardhat installation without external plugins. Also, there is no need to import `@nomicfoundation/hardhat-verify` as it is already included in `@nomicfoundation/hardhat-toolbox`.
